components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksInterceptor.kt

```diff
@@ -58,16 +58,17 @@ class AppLinksInterceptor(
         uri: String,
         hasUserGesture: Boolean,
         isSameDomain: Boolean,
-        isRedirect: Boolean
+        isRedirect: Boolean,
+        isDirectNavigation: Boolean
     ): RequestInterceptor.InterceptionResponse? {
         val uriScheme = Uri.parse(uri).scheme
         val engineSupportsScheme = engineSupportedSchemes.contains(uriScheme)
 
         val doNotIntercept = when {
             uriScheme == null -> true
-            // If request not from an user gesture and not from a redirect
+            // If request not from an user gesture, redirect and direct navigation
             // or if we're already on the site then let's not go to an external app.
-            ((!hasUserGesture && !isRedirect) || isSameDomain) && engineSupportsScheme -> true
+            ((!hasUserGesture && !isRedirect && !isDirectNavigation) || isSameDomain) && engineSupportsScheme -> true
```

Comment:
As a follow-up, I think we should try and refactor this and get rid of the many negations The conditions are very hard to read now. :)

Reply:
Sure, do you want me to update to `!(hasUserGesture || isRedirect || isDirectNavigation)` ?

Reply:
I don't think we have to do it right now, it will take some more work to make this easier to read, I think. We can lift `engineSupportsScheme` out of the block and check after. We could also remove the whole negation so instead of `doNotIntercept`, make it `shouldIntercept`, which should help with readability too, plus add lots of docs :).

Reply:
Sure.  Issue created to track this https://github.com/mozilla-mobile/android-components/issues/7614.  Thanks!